File: apps/mobile/modules/widget-bridge/ios/WidgetBridgeModule.swift
import ExpoModulesCore
import WidgetKit

/**
 A widget cannot run this app's JavaScript, so the only way data reaches one is
 shared native storage: the app writes, the widget reads (AGENTS.md §8).

 That store is an **App Group**, which both the app target and the widget
 extension must be entitled to. If the entitlement is missing the suite simply
 does not exist, so every call here reports failure rather than pretending to
 have written something the widget will never see.
 */
public class WidgetBridgeModule: Module {
  /// Must match the App Group in app.config.ts and the widget target.
  private static let appGroup = "group.de.mekteb.app"
  private static let payloadKey = "widgetPayload"

  public func definition() -> ModuleDefinition {
    Name("WidgetBridge")

    /// True when the App Group is reachable — the JS side uses this to decide
    /// whether widgets exist on this build at all.
    Function("isAvailable") { () -> Bool in
      UserDefaults(suiteName: WidgetBridgeModule.appGroup) != nil
    }

    /**
     Stores the payload as a JSON string, or clears it when `json` is nil.

     JSON rather than individual keys so the widget never renders half of one
     update and half of the next: one write, one read, one consistent picture.
     */
    AsyncFunction("setPayload") { (json: String?) -> Bool in
      guard let defaults = UserDefaults(suiteName: WidgetBridgeModule.appGroup) else {
        return false
      }

      if let json {
        defaults.set(json, forKey: WidgetBridgeModule.payloadKey)
      } else {
        defaults.removeObject(forKey: WidgetBridgeModule.payloadKey)
      }

      // Without this the widget keeps its last timeline until the OS decides
      // to refresh it, which can be hours.
      if #available(iOS 14.0, *) {
        WidgetCenter.shared.reloadAllTimelines()
      }
      return true
    }
  }
}

File: apps/mobile/modules/widget-bridge/android/src/main/java/expo/modules/widgetbridge/WidgetBridgeModule.kt
package expo.modules.widgetbridge

import android.content.Context
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * Android half of the widget data channel.
 *
 * A widget cannot run this app's JavaScript, so the only way data reaches one
 * is shared native storage: the app writes, the widget reads (AGENTS.md §8).
 *
 * On iOS the store is an App Group; on Android it is a plain
 * [SharedPreferences] file. The widget renders through
 * `react-native-android-widget`, whose headless task runs in the app's own
 * process, so `MODE_PRIVATE` on a fixed file name is sufficient — the task
 * reads exactly what the app wrote, as one JSON string.
 *
 * The module name must stay `WidgetBridge` — the JS side resolves
 * `requireOptionalNativeModule("WidgetBridge")` and expects the same shape on
 * both platforms.
 */
class WidgetBridgeModule : Module() {
  private companion object {
    /** SharedPreferences file, distinct from anything expo uses. */
    const val PREFS_NAME = "de.mekteb.app.widget"
    const val PAYLOAD_KEY = "widgetPayload"
  }

  private val reactContext: Context
    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()

  override fun definition() = ModuleDefinition {
    Name("WidgetBridge")

    /**
     * True when the store is reachable — which it always is on Android. The
     * JS side uses this to decide whether widgets exist on this build at all,
     * so it must not throw.
     */
    Function("isAvailable") {
      true
    }

    /**
     * Stores the payload as a JSON string, or clears it when `json` is null.
     *
     * JSON rather than individual keys so the widget never renders half of
     * one update and half of the next: one write, one read, one consistent
     * picture.
     */
    AsyncFunction("setPayload") { json: String? ->
      val prefs = reactContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
      if (json != null) {
        prefs.edit().putString(PAYLOAD_KEY, json).apply()
      } else {
        prefs.edit().remove(PAYLOAD_KEY).apply()
      }
      true
    }

    /**
     * Reads the payload the app last wrote. Called by the widget task handler
     * (headless JS) each time the OS asks the widget to render — see
     * `src/widgets/task-handler.ts`.
     */
    Function("getPayload") {
      reactContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        .getString(PAYLOAD_KEY, null)
    }
  }
}
